File: Client/ICQApp/ViewModels/AppViewModel.swift
import Foundation
import Combine

@MainActor
class AppViewModel: ObservableObject {
    @Published var isLoggedIn = false
    @Published var currentUser: UserDto?
    @Published var chats: [ChatDto] = []
    @Published var contacts: [UserDto] = []
    @Published var error: String?

    private let api = APIClient.shared

    init() {
        Task { await restoreSession() }
    }

    func restoreSession() async {
        await api.loadTokens()
        // try me() if token present
    }

    func login(emailOrUin: String, password: String) async {
        do {
            let resp = try await api.login(emailOrUin: emailOrUin, password: password)
            await api.setTokens(access: resp.accessToken, refresh: resp.refreshToken)
            currentUser = resp.user
            isLoggedIn = true
            await loadData()
        } catch {
            self.error = error.localizedDescription
        }
    }

    func register(email: String, password: String, nickname: String) async {
        do {
            let resp = try await api.register(email: email, password: password, nickname: nickname)
            await api.setTokens(access: resp.accessToken, refresh: resp.refreshToken)
            currentUser = resp.user
            isLoggedIn = true
            await loadData()
        } catch {
            self.error = error.localizedDescription
        }
    }

    func loadData() async {
        do {
            async let c = api.getChats()
            async let contacts = api.getContacts()
            self.chats = try await c
            self.contacts = try await contacts
        } catch {
            self.error = error.localizedDescription
        }
    }

    func logout() {
        Task { await api.setTokens(access: nil, refresh: nil) }
        isLoggedIn = false
        currentUser = nil
        chats = []
        contacts = []
    }
}

File: Android/app/src/main/java/com/icq/messenger/data/AppViewModel.kt
package com.icq.messenger.data

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.icq.messenger.network.ApiClient
import com.icq.messenger.network.SignalRClient
import kotlinx.coroutines.launch

class AppViewModel : ViewModel() {
    var isAuthenticated by mutableStateOf(false)
        private set
    var currentUser by mutableStateOf<UserDto?>(null)
        private set
    var chats by mutableStateOf<List<ChatDto>>(emptyList())
        private set
    var contacts by mutableStateOf<List<ContactDto>>(emptyList())
        private set
    var errorMessage by mutableStateOf<String?>(null)
    var isLoading by mutableStateOf(false)

    private val signalR = SignalRClient()

    fun login(nickname: String, password: String) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null
            try {
                val res = ApiClient.api.login(LoginRequest(nickname, password))
                handleAuth(res)
            } catch (e: Exception) {
                errorMessage = e.message ?: "Login failed"
            } finally {
                isLoading = false
            }
        }
    }

    fun register(nickname: String, password: String, email: String? = null) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null
            try {
                val res = ApiClient.api.register(RegisterRequest(nickname, password, email))
                handleAuth(res)
            } catch (e: Exception) {
                errorMessage = e.message ?: "Register failed"
            } finally {
                isLoading = false
            }
        }
    }

    fun logout() {
        signalR.disconnect()
        ApiClient.accessToken = null
        isAuthenticated = false
        currentUser = null
        chats = emptyList()
        contacts = emptyList()
    }

    private fun handleAuth(res: AuthResponse) {
        ApiClient.accessToken = res.accessToken
        currentUser = res.user
        isAuthenticated = true
        signalR.connect(res.accessToken)
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            try {
                chats = ApiClient.api.getChats()
                contacts = ApiClient.api.getContacts()
            } catch (e: Exception) {
                errorMessage = e.message
            }
        }
    }

    fun startChat(userId: String, onCreated: (ChatDto) -> Unit) {
        viewModelScope.launch {
            try {
                val chat = ApiClient.api.createPrivateChat(CreatePrivateChatRequest(userId))
                if (chats.none { it.id == chat.id }) {
                    chats = listOf(chat) + chats
                }
                onCreated(chat)
            } catch (e: Exception) {
                errorMessage = e.message
            }
        }
    }

    fun signalRClient(): SignalRClient = signalR
}
